### PR TITLE
Add a feature based object store provider

### LINKED ISSUE #257: Add a feature based object store provider
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently ObjectStore cannot be self-registered by its path url, which is very important feature for Ballista to support remote storage, like HDFS, etc. 

**Describe the solution you'd like**

By introduce *FeatureBasedObjectStoreProvider*, we can just add features for different object stores for the Ballista. And then we can detect which object store to be used by URL without do any manual object store registration.

**Describe alternatives you've considered**

**Additional context**

https://github.com/apache/arrow-datafusion/issues/2905

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #257, #6.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @andygrove, @tustvold, @matthewmturner, could you help review this PR?

**Review by tustvold — COMMENTED** [review]
I don't really have enough context on ballista to meaningfully review this, although its use of ObjectStoreProvider looks plausible.

https://github.com/apache/arrow-datafusion/pull/3540 might also be related, a common way to register object stores based on environment variables seems sensible to me

**Comment by andygrove** [comment]
cc @avantgardnerio who has been working on supporting custom object stores in DataFusion/Ballista

**Comment by andygrove** [comment]
The approach here LGTM but I have not worked with the object store yet

File: ballista/rust/core/src/utils.rs [hunk 1]
```diff
@@ -54,9 +56,38 @@ use std::time::Duration;
...
+
+/// Get a RuntimeConfig with specific ObjectStoreDetector in the ObjectStoreRegistry
+pub fn with_object_store_provider(config: RuntimeConfig) -> RuntimeConfig {
+    config.with_object_store_registry(Arc::new(ObjectStoreRegistry::new_with_provider(
+        Some(Arc::new(FeatureBasedObjectStoreProvider)),
+    )))
+}
 
+/// An object store detector based on which features are enable for different kinds of object stores
+pub struct FeatureBasedObjectStoreProvider;
```
[the remark this answers is not in the document]
Comment (andygrove) [reply ballista/rust/core/src/utils.rs:80]: There is a follow-on PR https://github.com/apache/arrow-ballista/pull/260 to implement this for HDFS but it is waiting on another PR to be merged

[the remark this answers is not in the document]
Comment (yahoNanJing) [reply ballista/rust/core/src/utils.rs:80]: Yes, based on this PR, we can easily add feature-based object stores like #260 does. And we have one verified example in https://github.com/datafusion-contrib/datafusion-objectstore-hdfs/blob/master/hdfs-examples/src/bin/ballista-sql.rs

**Comment by yahoNanJing** [comment]
Let's merge it first. Will update the *ObjectStoreProvider* interface after upgrading datafusion to the latest version.

**Comment by avantgardnerio** [comment]
It's unclear from skimming this: what is the intended usage? Should a statement like:

```
create external table customer stored as CSV DELIMITER '|' location 's3://minioadmin:minioadmin@127.0.0.1:9000/tpch';
```

Work now?

**Pull request merged**